### ISSUE TITLE
MappingValidationException when using abstract class

### ISSUE THREAD

**Issue by shaneh20** [agileobjects/AgileMapper#183]
Hi, my code below seems to throw a MappingValidationException when using Abstract base classes with the Mapper.WhenMapping.ThrowIfAnyMappingPlanIsIncomplete setting 'turned on'.

```
Imports AgileObjects.AgileMapper

Module Module1

    Sub Main()
        'Uncomment/comment the following to cause/prevent exception when Mapper.Map is called below:
        Mapper.WhenMapping.ThrowIfAnyMappingPlanIsIncomplete()
        Mapper.WhenMapping.From(Of ThingBase).To(Of ThingBaseDTO)()
        Mapper.WhenMapping.From(Of Container).To(Of ContainerDTO)()

        Dim container As New Container
        Dim containerDTO = Mapper.Map(container).ToANew(Of ContainerDTO)
    End Sub

End Module

Public MustInherit Class ThingBase

End Class

Public Class Thing
    Inherits ThingBase

End Class

Public MustInherit Class ThingBaseDTO

End Class

Public Class ThingDTO
    Inherits ThingBaseDTO

End Class

Public Class Container

    Public Property Something As ThingBase

End Class

Public Class ContainerDTO

    Public Property Something As ThingBaseDTO

End Class
```

The exception contains the message:

> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
> - Container -> ContainerDTO
> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
> 
>  Rule set: CreateNew
> 
>  Unconstructable target Types - fix by ignoring or configuring constructor parameters:
> 
>   - ThingBase -> ThingBaseDTO

I'm new to Agile Mapper, so I'm not sure if this exception is to be expected, or if I'm doing something stupid. I would have thought my mappings were valid, so I'm thinking that this exception should not be throwing. 

Is there are way I can prevent this exception without having to 'turn off' the Mapper.WhenMapping.ThrowIfAnyMappingPlanIsIncomplete setting?

Thanks

**Comment by SteveWilkes** [comment]
Hi!

Thanks for the question :) The mapper is complaining because it doesn't know what to do when mapping from `Container.ThingBase` -> `ContainerDTO.ThingBaseDTO` as `ThingBaseDTO` is abstract and we haven't told it what to create when mapping one.

Depending on your expectations, this may or may not be a bug - [without](https://dotnetfiddle.net/7puml0) `ThrowIfAnyMappingPlanIsIncomplete`, the mapper will discover and pair up `Thing` and `ThingDTO` at runtime, but given there could be other derived types it can't find or pair up, is that a complete mapping plan? I guess seeing as it can find at least one derived type pair, it shouldn't throw the exception. I'll fix that!

Thanks,

Steve

**Comment by SteveWilkes** [comment]
So here's where I've got to with this one so far:

Lets say `Thing` and `ThingDTO` both have a `Value` member. Mapping from `ThingBase` to `ThingBaseDTO` produces several potential mappings:

| c.Something | cDTO.Something | Source type | Result type | ThingDTO.Value source |
|---|---|---|---|---|
| `Thing` | null | `Thing` | `ThingDTO` | `Thing.Value` |
| `Thing` | `ThingBaseDTO` | `Thing` | `ThingBaseDTO` | `Thing.Value` |
| `Thing` | `ThingDTO` | `Thing` | `ThingDTO` | `Thing.Value` |
| `ThingBase` | null | `ThingBase` | null | N/A |
| `ThingBase` | `ThingBaseDTO` | `ThingBase` | `ThingBaseDTO` | N/A |
| `ThingBase` | `ThingDTO` | `ThingBase` | `ThingDTO` | ?! |

The problem comes mapping `ThingBase` (so a non-null source which isn't a `Thing`) to an existing `ThingDTO` target; Mapping validation fails as as we have nothing to map to `ThingDTO.Value`. We can't fix this by ignoring the target member:

```csharp
mapper.WhenMapping
    .From<ThingBase>()
    .To<ThingDTO>()
    .Ignore(dto => dto.Value);
```

...as that also ignores `ThingDTO.Value` when mapping from any source types _derived_ from `ThingBase`, which we don't want.

So I'm working on this:

```csharp
mapper.WhenMapping
    .From<ThingBase>().ButNotDerivedTypes
    .To<ThingDTO>()
    .Ignore(dto => dto.Value);
```

...which will only apply the ignore to `ThingBase`, not `Thing` or any other derived types.

**Comment by shaneh20** [comment]
I'm not sure I completely understand what you are saying (my brain is still a bit cooked from new years), but I think that would be ok.

Although I don't _think_ this problem would affect me as I doubt I would ever have models and DTO's defined as Container.Something (ThingBase) and ContainerDTO.Something (ThingDTO). They would always be defined as Container.Something (ThingBase) and ContainerDTO.Something (ThingBaseDTO). 

So, from my perspective it would be good enough if Agile Mapper detected when Container.Something and ContainerDTO.Something used _mismatching_ types and threw a validation error. I guess Agile Mapper would need to cater for all other more complex scenarios though.

**Comment by SteveWilkes** [comment]
This is fixed / supported in the latest v1.6 preview branch code - release to follow.

Thanks again for the feedback!

Steve

**Comment by SteveWilkes** [comment]
This fix is include in v1.6, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper).

Thanks again for the feedback!
